bundles/framework/divmanazer/component/Chart.js

```diff
@@ -57,25 +56,85 @@ Oskari.clazz.define('Oskari.userinterface.component.Chart', function() {
         // TODO: we need some proper handling for missing values AND negative values.
-        this.x = d3.scaleLinear()
-        .domain([ 0, d3.max( this.data, function ( d ) {
-                return d.value;
-            })
-        ]);
 
-        this.y = d3.scaleBand()
-        .domain( this.data.map ( function ( d ) {
+        this.x = d3.scaleLinear();
+        this.y = d3.scaleBand();
+        var dataset = this.getDatasetMinMax();
+        var xScaleDomain;
+
+        if ( !this.dataHasNegativeValues() ) {
+            xScaleDomain = [ 0, dataset.max]
+        } else {
+            xScaleDomain = d3.extent(this.data, function ( d ) {return d.value; } );
+        }
+        var yScaleDomain = this.data.map( function ( d ) {
             return d.name;
-        }));
+        });
+        this.x.domain( xScaleDomain );
+        this.y.domain( yScaleDomain );
     },
     getSVGTemplate: function () {
-        var svg = d3.select( this.graph.get(0) ).append("svg")
+        var svg = d3.select( this.plot.get(0) ).append("svg")
             .attr( "width", this.dimensions.width() + this.dimensions.margin.left + this.dimensions.margin.right )
             .attr( "height", this.dimensions.height() + this.dimensions.margin.top + this.dimensions.margin.bottom )
             .append( "g" )
             .attr( "transform", "translate(" + this.dimensions.margin.left + "," + this.dimensions.margin.top + ")" );
+
         return svg;
     },
+    getDatasetMinMax: function () {
+        var min = d3.min(this.data, function ( d ) { return d.value; });
+        var max = d3.max(this.data, function ( d ) { return d.value; });
+        return {
+            min: min,
+            max: max
+        }
+    },
+    /**
+     *
+     * @method sortDataByType
+     * @param { String } type - indicates what sort of sorting we want to apply to our chart.
+     * if no type provided it will default to value ascending
+     */
+    sortDataByType: function ( type ) {
+        var me = this;
+        // empty string ("") is the placeholder value from selectlist
+        if ( !this.sortingType ) {
+            if ( typeof type === "undefined" || type === "" ) {
+                type = 'value-descending';
+            } 
+        } else if( type === undefined || type === "" ) {
+            type = this.sortingType;
+        }
+
+        var sortTypes = {
+            "name-ascending": function () {
+                me.data.sort(function (a, b) {
+                    return d3.descending(a.name , b.name);
+                });
+            },
+            "name-descending": function () {
+                me.data.sort(function (a, b) {
+                    return d3.ascending(a.name , b.name);
+                });
+            },
+            "value-ascending": function () {
+                me.data.sort(function (a, b) {
+                    return d3.descending(a.value || 1000000 , b.value || 1000000);
```

Comment:
What if `a.value` or `b.value` is number zero? Some values might be over 1000000, too. 

We should do an early return from the comparator instead: 
```
if(a.value == null && b.value == null) return 0;
if(a.value == null) return -1;
if(b.value == null) return 1;
return d3.descending(a.value, b.value);
```

While at it we could refactor the logic into a `nullsLast()` comparator, that we could then call in other places like so:
```
var result = nullsLast(a.value, b.value);
if(result === 0) {
    return d3.descending(a.value, b.value);
}
return result;

```